src/init.ml

```diff
@@ -0,0 +1,198 @@
+)
+" C.pname C.pname
+    in
+    make_file name content
+
+  let lib_ml =
+    let name = "lib.ml" in
+    let content = {|let hello name = Printf.sprintf "Hello, %s" name
+|} in
+    make_file name content
+
+  let files = [lib_dune; lib_ml]
+end
+
+module Bin (C : InitContext) : ProjectDir = struct
+  let dir = Path.(append C.dir (of_string "bin"))
+  let make_file = File.make dir
+
+  let bin_dune =
+    let name = "dune" in
+    let content = Printf.sprintf "\
+(executable
+  (public_name %s)
+  (name main)
+  (libraries %s))
+" C.pname C.pname
+    in
+    make_file name content
+
+  let bin_ml =
+    (* The depreciated [String.capitalize] is required over the preferred
+       [String.capitalize_ascii] to support OCaml 4.02.3 *)
+    let lib_module_name = String.capitalize C.pname in
+    let name = "main.ml" in
+    let content =
+      Printf.sprintf
+        {|let () = print_endline (%s.Lib.hello "project %s")
+|}
+        lib_module_name C.pname
+    in
+    make_file name content
+
+  let files = [bin_dune; bin_ml]
+end
+
+let make_project_dir_exn (module D : ProjectDir) =
+  if Path.exists D.dir then
+    Exn.fatalf "A project directory cannot be created: \
+                A directory named %s already exists"
+      (Path.to_string D.dir);
+
+  try Path.mkdir_p D.dir with
+  | Unix.Unix_error (EACCES, _, _) ->
+    Exn.fatalf "A project directory cannot be created: \
+                Lacking permissions needed to create directory %s"
+      (Path.to_string D.dir)
+
+let make_project_dir_files (module D : ProjectDir) =
+  List.iter ~f:File.create D.files
+
+let validate_project_name name =
+  match Lib_name.Local.of_string name with
+  | Ok _ -> ()
+  | _ -> Exn.fatalf "A project named '%s' cannot be created because it is an %s"
+           name Lib_name.Local.invalid_message
+
+let make_context pname kind path =
+  let module C = struct
+    let kind = kind
+    let pname = pname
+    let dir =
+      match path with
+      | None -> Path.of_filename_relative_to_initial_cwd pname
+      | Some path ->
+        if Filename.is_relative path
+        then Path.of_filename_relative_to_initial_cwd path
+        else Path.of_string path
+  end
+  in
+  (module C : InitContext)
+
+let make_project_dirs (module C : InitContext) =
+  (* dirs that are always included in a project *)
+  let project_dirs : (module ProjectDir) list =
+    [
+      (module Proj (C)); (* The project dir should always be first *)
+      (module Lib (C));
+    ]
+  in
+  let project_dirs = match C.kind with
+    | Project.Kind.Library -> project_dirs
+    | Project.Kind.Executable -> project_dirs @ [(module Bin (C))]
+  in
+  List.iter ~f:make_project_dir_exn project_dirs;
+  List.iter ~f:make_project_dir_files project_dirs
+
+let print_completion (module C : InitContext) =
+  Errors.kerrf ~f:print_to_console
+    "@{<ok>Success@}: %s %s project initialized\n"
+    C.pname (Project.Kind.to_string C.kind)
+
+let print_hints (module C : InitContext) =
```

Comment:
IMO, these hints are annoying for non new users. If the feature is aimed at experienced users as well, we should avoid them.

Reply:
These particular hints are less annoying than the commented lines btw. As they don't require manual removal. If you'd prefer, you can keep these but remove the commented stuff.

The library and executable stanzas should already have the  key options exposed via `$ dune init --help` and down the road, once we have auto completion for dune files, then those commented lines will be even less useful.